File: exercise/kotlin/day13/src/main/kotlin/santamarket/model/Discount.kt
package santamarket.model

import kotlin.math.abs

data class Discount(
    val product: Product,
    val description: String,
    val discountAmount: Double
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Discount) return false

        val tolerance = 0.001
        return product == other.product &&
                description == other.description &&
                abs(discountAmount - other.discountAmount) < tolerance
    }

    override fun hashCode(): Int = arrayOf(product, description, discountAmount).contentHashCode()
}

File: exercise/scala/day13/src/main/scala/santamarket/model/Discount.scala
package santamarket.model

case class Discount(product: Product, description: String, discountAmount: Double):
  override def equals(obj: Any): Boolean = obj match
    case that: Discount =>
      this.description == that.description &&
        Math.abs(this.discountAmount - that.discountAmount) < 0.001 &&
        this.product == that.product
    case _ => false

  override def hashCode(): Int =
    (description, discountAmount, product).hashCode